projects/geometry/110_mesh_shader_triangle_d3d12/110_mesh_shader_triangle_d3d12.cpp

```diff
@@ -0,0 +1,330 @@
+// Globals
+// =============================================================================
+static uint32_t gWindowWidth  = 1280;
+static uint32_t gWindowHeight = 720;
+static bool     gEnableDebug  = true;
+
+static LPCWSTR gVSShaderName = L"vsmain";
+static LPCWSTR gPSShaderName = L"psmain";
+
+void CreateGlobalRootSig(DxRenderer* pRenderer, ID3D12RootSignature** ppRootSig);
+void CreateGeometryBuffers(
+    DxRenderer*      pRenderer,
+    ID3D12Resource** ppIndexBuffer,
+    ID3D12Resource** ppVertexBuffer,
+    ID3D12Resource** ppVertexColorBuffer);
+
+// =============================================================================
+// main()
+// =============================================================================
+int main(int argc, char** argv)
+{
+    std::unique_ptr<DxRenderer> renderer = std::make_unique<DxRenderer>();
+
+    if (!InitDx(renderer.get(), gEnableDebug))
+    {
+        return EXIT_FAILURE;
+    }
+
+    D3D12_FEATURE_DATA_D3D12_OPTIONS7 options7 = {};
+    //
+    CHECK_CALL(renderer->Device->CheckFeatureSupport(D3D12_FEATURE_D3D12_OPTIONS7, &options7, sizeof(options7)));
+
+    bool isMeshShadingSupported = (options7.MeshShaderTier >= D3D12_MESH_SHADER_TIER_1);
+    if (!isMeshShadingSupported)
+    {
+        assert(false && "Required mesh shading tier not supported");
+        return EXIT_FAILURE;
+    }
+
+    // *************************************************************************
+    // Compile shaders
+    // *************************************************************************
+    std::vector<char> dxilMS;
+    std::vector<char> dxilPS;
+    {
+        std::string errorMsg;
+        HRESULT     hr = CompileHLSL(gShaders, "msmain", "ms_6_5", &dxilMS, &errorMsg);
+        if (FAILED(hr))
+        {
+            std::stringstream ss;
+            ss << "\n"
+               << "Shader compiler error (MS): " << errorMsg << "\n";
+            GREX_LOG_ERROR(ss.str().c_str());
+            assert(false);
+            return EXIT_FAILURE;
+        }
+
+        hr = CompileHLSL(gShaders, "psmain", "ps_6_5", &dxilPS, &errorMsg);
+        if (FAILED(hr))
+        {
+            std::stringstream ss;
+            ss << "\n"
+               << "Shader compiler error (PS): " << errorMsg << "\n";
+            GREX_LOG_ERROR(ss.str().c_str());
+            assert(false);
+            return EXIT_FAILURE;
+        }
+    }
+
+    // *************************************************************************
+    // Root signature
+    // *************************************************************************
+    ComPtr<ID3D12RootSignature> rootSig;
+    CreateGlobalRootSig(renderer.get(), &rootSig);
+
+    // *************************************************************************
+    // Graphics pipeline state object
+    // *************************************************************************
+    D3DX12_MESH_SHADER_PIPELINE_STATE_DESC psoDesc           = {};
```

Comment:
You don't want to stick all this behind a CreateMeshShaderPipeline() like we did for other examples?

Reply:
I wantd to keep it in here so that it's not buried somewhere in the renderer code. I couldn't find any other sample on the internet that shows a simple straight forward pipeline creation so  I figured this might be helpful.